### ISSUE TITLE
container "pytorch" is waiting to start: PodInitializing

### ISSUE THREAD

**Issue by gogogwwb** [kubeflow/pytorch-operator#348]
When the master is finished running, the worker is still initializing.

![1629019887(1)](https://user-images.githubusercontent.com/51263129/129474058-22d7190f-e8e6-4ff3-a623-ed5d1dbcd7cd.jpg)

worker log：
Error from server (BadRequest): container "pytorch" in pod "jupyter-kernel-jxosi-worker-0" is waiting to start: PodInitializing

What is the reason for this?

**Comment by gaocegege** [comment]
Could you please run `kubectl describe` and post the result here?

**Comment by gogogwwb** [comment]
![image](https://user-images.githubusercontent.com/51263129/129506361-a2ae4a04-74da-457a-85b5-9714d8079a93.png)

**Comment by gaocegege** [comment]
Can you show more about it? Especially the events section.

**Comment by gogogwwb** [comment]
![image](https://user-images.githubusercontent.com/51263129/129513646-69cd9455-ce53-471a-b167-2ba3f86e264c.png)

**Comment by gaocegege** [comment]
Seems that the init container is pending. Can you show its log?

**Comment by gogogwwb** [comment]
init-pytorch log：
![image](https://user-images.githubusercontent.com/51263129/129522859-2ea8cab2-975d-4c27-9ef2-70b4fa23aa5f.png)

**Comment by gogogwwb** [comment]
master svc ：

![image](https://user-images.githubusercontent.com/51263129/129523423-5a370492-2fbd-4e8b-ac8c-264245ad991f.png)

**Comment by gaocegege** [comment]
Can you try kubectl debug to run an ephemeral container, then run `ping xxx-master-0`?

**Comment by gogogwwb** [comment]
ping master-0 in an ephemeral container :
![image](https://user-images.githubusercontent.com/51263129/129564642-8b382f3a-8f4e-47e5-8aba-8200db387ecd.png)

**Comment by gogogwwb** [comment]
I found ep appeared, but disappeared again after a while
![image](https://user-images.githubusercontent.com/51263129/129565103-3cf34da7-15d9-4e19-952c-3a09374967be.png)
![image](https://user-images.githubusercontent.com/51263129/129565227-cb73d49b-4970-401e-bea1-7d84cbe6665a.png)

**Comment by gaocegege** [comment]
It's weird.

**Comment by gogogwwb** [comment]
I put the program to sleep for a while and found that the worker can run. Is there any restriction on the creation order of service and pod in pytorchjob?

**Comment by gogogwwb** [comment]
It should be that the master executes too fast, so that the ep of the service finally becomes none, and the worker cannot obtain the IP address of the master.

**Comment by gaocegege** [comment]
> It should be that the master executes too fast, so that the ep of the service finally becomes none, and the worker cannot obtain the IP address of the master.

Interesting. /cc @johnugeorge

**Comment by johnugeorge** [comment]
But in that case, master should not start the job until workers are up. Are you using distributed setup itself in the code?

**Comment by gogogwwb** [comment]
I did not use distributed steps in the code. After master running, it becomes completed state, "kubectl get ep -ntest", and found that ep is none

**Comment by gogogwwb** [comment]
> But in that case, master should not start the job until workers are up. Are you using distributed setup itself in the code?

If the code is not sleeping, the master is in a short running state, but the worker is in the init state

**Comment by johnugeorge** [comment]
If you are not using distributed pytorch in the code, this can happen.  Master can start executing and gets completed before worker starts. Can you confirm whether you are using distributed APIs?

**Comment by gogogwwb** [comment]
import torch
import time
import os
package_name = 'requests'
os.system(f'pip install {package_name}')
dtype = torch.float
device = torch.device("cpu")
# device = torch.device("cuda:0") # Uncomment this to run on GPU
#time.sleep(60)
# N is batch size; D_in is input dimension;
# H is hidden dimension; D_out is output dimension.
N, D_in, H, D_out = 64, 1000, 100, 10

# Create random input and output data
x = torch.randn(N, D_in, device=device, dtype=dtype)
y = torch.randn(N, D_out, device=device, dtype=dtype)

# Randomly initialize weights
w1 = torch.randn(D_in, H, device=device, dtype=dtype)
w2 = torch.randn(H, D_out, device=device, dtype=dtype)
print(torch.__version__)

learning_rate = 1e-6
for t in range(500):
    # Forward pass: compute predicted y
    h = x.mm(w1)
    h_relu = h.clamp(min=0)
    y_pred = h_relu.mm(w2)

    # Compute and print loss
    loss = (y_pred - y).pow(2).sum().item()
    print(t, loss)

    # Backprop to compute gradients of w1 and w2 with respect to loss
    grad_y_pred = 2.0 * (y_pred - y)
    grad_w2 = h_relu.t().mm(grad_y_pred)
    grad_h_relu = grad_y_pred.mm(w2.t())
    grad_h = grad_h_relu.clone()
    grad_h[h < 0] = 0
    grad_w1 = x.t().mm(grad_h)

    # Update weights using gradient descent
    w1 -= learning_rate * grad_w1
    w2 -= learning_rate * grad_w2

**Comment by gogogwwb** [comment]
import torch
import time
import os
package_name = 'requests'
os.system(f'pip install {package_name}')
dtype = torch.float
device = torch.device("cpu")
# device = torch.device("cuda:0") # Uncomment this to run on GPU
#time.sleep(60)
# N is batch size; D_in is input dimension;
# H is hidden dimension; D_out is output dimension.
N, D_in, H, D_out = 64, 1000, 100, 10

# Create random input and output data
x = torch.randn(N, D_in, device=device, dtype=dtype)
y = torch.randn(N, D_out, device=device, dtype=dtype)

# Randomly initialize weights
w1 = torch.randn(D_in, H, device=device, dtype=dtype)
w2 = torch.randn(H, D_out, device=device, dtype=dtype)
print(torch.__version__)

learning_rate = 1e-6
for t in range(500):
    # Forward pass: compute predicted y
    h = x.mm(w1)
    h_relu = h.clamp(min=0)
    y_pred = h_relu.mm(w2)

    # Compute and print loss
    loss = (y_pred - y).pow(2).sum().item()
    print(t, loss)

    # Backprop to compute gradients of w1 and w2 with respect to loss
    grad_y_pred = 2.0 * (y_pred - y)
    grad_w2 = h_relu.t().mm(grad_y_pred)
    grad_h_relu = grad_y_pred.mm(w2.t())
    grad_h = grad_h_relu.clone()
    grad_h[h < 0] = 0
    grad_w1 = x.t().mm(grad_h)

    # Update weights using gradient descent
    w1 -= learning_rate * grad_w1
    w2 -= learning_rate * grad_w2

**Comment by gogogwwb** [comment]
`import torch
import time
import os
package_name = 'requests'
os.system(f'pip install {package_name}')
dtype = torch.float
device = torch.device("cpu")
# device = torch.device("cuda:0") # Uncomment this to run on GPU
#time.sleep(60)
# N is batch size; D_in is input dimension;
# H is hidden dimension; D_out is output dimension.
N, D_in, H, D_out = 64, 1000, 100, 10

# Create random input and output data
x = torch.randn(N, D_in, device=device, dtype=dtype)
y = torch.randn(N, D_out, device=device, dtype=dtype)

# Randomly initialize weights
w1 = torch.randn(D_in, H, device=device, dtype=dtype)
w2 = torch.randn(H, D_out, device=device, dtype=dtype)
print(torch.__version__)

learning_rate = 1e-6
for t in range(500):
    # Forward pass: compute predicted y
    h = x.mm(w1)
    h_relu = h.clamp(min=0)
    y_pred = h_relu.mm(w2)

    # Compute and print loss
    loss = (y_pred - y).pow(2).sum().item()
    print(t, loss)

    # Backprop to compute gradients of w1 and w2 with respect to loss
    grad_y_pred = 2.0 * (y_pred - y)
    grad_w2 = h_relu.t().mm(grad_y_pred)
    grad_h_relu = grad_y_pred.mm(w2.t())
    grad_h = grad_h_relu.clone()
    grad_h[h < 0] = 0
    grad_w1 = x.t().mm(grad_h)

    # Update weights using gradient descent
    w1 -= learning_rate * grad_w1
    w2 -= learning_rate * grad_w2`

**Comment by gogogwwb** [comment]
![image](https://user-images.githubusercontent.com/51263129/133072607-6c2862f8-eab1-475e-abc4-e2323830f900.png)

**Comment by johnugeorge** [comment]
That is issue. Any reason in using pytorch-operator without using distributed version?

Example:

https://github.com/kubeflow/tf-operator/blob/1aa44a68cd364ed6e30c0841e6daf1d93a29f146/examples/pytorch/mnist/mnist.py#L72

**Comment by gogogwwb** [comment]
How do I make the pod created by pytorchjob not automatically disappear after completion, is it to use cleanpolicy? How to set it up? thanks

**Comment by johnugeorge** [comment]
Set to None

https://github.com/kubeflow/common/blob/f7c41a08761ff3b215553a051fd529efd22782a1/pkg/apis/common/v1/types.go#L136
